plugins/artifacts/artifacts.go

```diff
@@ -0,0 +1,110 @@
+//go:generate go-composite-schema --unexported --required artifacts config-schema.yml generated_configschema.go
+
+// Package artifacts is responsible for uploading artifacts after builds
+package artifacts
+
+import (
+	"fmt"
+	"mime"
+	"path/filepath"
+
+	"github.com/taskcluster/taskcluster-client-go/tcclient"
+	"github.com/taskcluster/taskcluster-worker/engines"
+	"github.com/taskcluster/taskcluster-worker/plugins"
+	"github.com/taskcluster/taskcluster-worker/plugins/extpoints"
+	"github.com/taskcluster/taskcluster-worker/runtime"
+	"github.com/taskcluster/taskcluster-worker/runtime/ioext"
+)
+
+type pluginProvider struct {
+	extpoints.PluginProviderBase
+}
+
+func (pluginProvider) NewPlugin(extpoints.PluginOptions) (plugins.Plugin, error) {
+	return plugin{}, nil
+}
+
+func (pluginProvider) ConfigSchema() runtime.CompositeSchema {
+	return runtime.NewEmptyCompositeSchema()
+}
+
+type plugin struct {
+	plugins.PluginBase
+}
+
+func (plugin) PayloadSchema() (runtime.CompositeSchema, error) {
+	return configSchema, nil
+}
+
+func (plugin) NewTaskPlugin(options plugins.TaskPluginOptions) (plugins.TaskPlugin, error) {
+	return &taskPlugin{
+		TaskPluginBase: plugins.TaskPluginBase{},
+		payload:        *(options.Payload.(*config)),
+	}, nil
+}
+
+type taskPlugin struct {
+	plugins.TaskPluginBase
+	context *runtime.TaskContext
+	payload config
+}
+
+func (tp *taskPlugin) Prepare(context *runtime.TaskContext) error {
+	tp.context = context
+	return nil
+}
+
+func (tp *taskPlugin) Stopped(result engines.ResultSet) (bool, error) {
+	for _, artifact := range tp.payload {
+		switch artifact.Type {
+		case "directory":
+			err := result.ExtractFolder(artifact.Path, tp.createUploadHandler(artifact.Name, artifact.Expires))
+			if err != nil {
+				runtime.CreateErrorArtifact(runtime.ErrorArtifact{
+					Message: fmt.Sprintf("Could not open directory '%s'", artifact.Path),
+					Reason:  "invalid-resource-on-worker",
+					Expires: artifact.Expires,
+				}, tp.context)
+			}
+		case "file":
+			fileReader, err := result.ExtractFile(artifact.Path)
+			if err != nil {
+				runtime.CreateErrorArtifact(runtime.ErrorArtifact{
+					Message: fmt.Sprintf("Could not read file '%s'", artifact.Path),
+					Reason:  "file-missing-on-worker",
+					Expires: artifact.Expires,
+				}, tp.context)
+			}
+			tp.attemptUpload(fileReader, artifact.Path, artifact.Name, artifact.Expires)
+		}
+	}
+	// TODO: Don't always return true?
+	return true, nil
```

Comment:
If there was errors uploading artifacts (and we used all retries)... we return the http error... or wrap it... this will be a fatal error, no way around... Maybe later we'll make that an internal error.

If there was an illegal filename, feature not supported, MalformedPayloadError, then we create a MalformedPayloadError explaining what when wrong...

In the face of: ErrResourceNotFound
We return `false, nil`, and we print to log that the task failed because the file or folder was missing.